[Merged by Bors] - feat(linear_algebra/basic): add missing lemma finsupp.sum_smul_index_linear_map'

ocfnash opened the pull request:
See also [this Zulip conversation](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there.20code.20for.20X.3F/topic/Sum.20is.20linear/near/229021943). cc @eric-wieser 

Co-authored-by: Adam Topaz <github@adamtopaz.com>
Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

---
Note that the similar-looking (but different!) lemma `smul_sum` is directly above (on line 70). I fully admit that `smul_sum'` is not a great name. I couldn't think of anything I liked better but suggestions are welcome.

ocfnash committed:
feat(linear_algebra/basic): add missing lemma finsupp.smul_sum'

eric-wieser commented on src/linear_algebra/basic.lean:
```diff
@@ -73,6 +73,19 @@ lemma smul_sum {α : Type u} {β : Type v} {R : Type w} {M : Type y}
+  { intros a m v ha hm ih,
+    simp only [sum_add_index, ih, smul_add, linear_map.map_zero, implies_true_iff, eq_self_iff_true,
+      smul_single, forall_3_true_iff, sum_single_index, linear_map.map_smul, linear_map.map_add], },
+end
```
Here's a shorter proof:

```suggestion
begin
  rw [finsupp.sum_smul_index', finsupp.smul_sum],
  { simp only [linear_map.map_smul], },
  { intro i, exact (h i).map_zero },
end
```

ocfnash replied:
🍾

ocfnash committed:
Update src/linear_algebra/basic.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ocfnash committed:
Add Eric's `finsupp.sum_smul_index_add_monoid_hom`

eric-wieser commented on src/linear_algebra/basic.lean:
```diff
@@ -73,6 +73,17 @@ lemma smul_sum {α : Type u} {β : Type v} {R : Type w} {M : Type y}
 finset.smul_sum
 
+@[simp]
+lemma smul_sum' {α : Type u} {R : Type v} {M : Type w} {M₂ : Type x}
```
This name should probably be `sum_smul_index_linear_map'` or similar, since the LHS is the same as `sum_smul_index`

jcommelin replied:
Maybe just `sum_smul_index'`?

ocfnash replied:
I just added a commit renaming to `sum_smul_index_linear_map'` since `sum_smul_index'` is already taken.

jcommelin reviewed (commented):
Ok, let me suggest that we just merge this as it is. New lemmas can be added in later PRs.

ocfnash committed:
Rename `smul_smul'` → `sum_smul_index_linear_map'`

I also liked Johan's suggestion `sum_smul_index'` but unfortunately it is
already taken.
